pkg/controller/garbagecollector/metrics/metrics.go

```diff
@@ -0,0 +1,46 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package metrics
+
+import (
+	"sync"
+
+	"k8s.io/component-base/metrics"
+	"k8s.io/component-base/metrics/legacyregistry"
+)
+
+const GarbageCollectorControllerSubsystem = "garbagecollector_controller"
+
+var (
+	GarbageCollectorResourcesSyncError = metrics.NewCounter(
+		&metrics.CounterOpts{
+			Subsystem:      GarbageCollectorControllerSubsystem,
+			Name:           "garbagecollector_resources_sync_error_total",
```

Comment:
Seeing how the metric is used, there seem to be different reasons behind sync failures. As such, would it make sense to introduce a `reason` label to differentiate them more easily?

Reply:
I suppose it could be somewhat useful, but usually you want to go to see the logs once the error occurs.

Reply:
you'd have to be careful with cardinality if you go with this route though, otherwise this can cause memory issues.

Reply:
There are two possible values at the moment and I am not expecting them to increase, so I hope that is ok

Reply:
I suggest explicitly bounding the values in a conditional and default to 'unknown' for any other ones.

Reply:
I have done exactly that by introducing RecordGarbageCollectorSyncErrorMetric function.

But I have run into a problem. By introducing reason label value, I am not picking up any metric values before the first error is recorded. I played with query quite a bit but I am unable to register a change from 0 -> 1 error which is essentially null -> 1 and isn't picked up by `rate`.


Query I am trying to achieve for an alert: `rate(garbagecollector_controller_garbagecollector_resources_sync_error_total{}[1h]) > 0`

I am seeing two options:

1. force initialize metric for all reasons with 0 at the begining.
2. revert back to not using labels which would make observability simpler. As I said before, the benefit of adding the reason is not that big.

thougths?

Reply:
> I suppose it could be somewhat useful, but usually you want to go to see the logs once the error occurs.

If you don't have any use for the reason label in terms of alerting, I think it is correct to stick with the approach of looking into the logs when an error occurs, although it might be simpler to investigate if you have the reason in the metric.

> But I have run into a problem. By introducing reason label value, I am not picking up any metric values before the first error is recorded. I played with query quite a bit but I am unable to register a change from 0 -> 1 error, which is essentially null -> 1 and isn't picked up by rate.

Not seeing the error metric until an actual error occurs is a normal monitoring behavior. For example, the apiserver request metric will only show metrics with a 500 status code if the apiserver answers a request with 500.

Going from 0 -> 1 to null -> 1 should be the same in Prometheus, so there might have been something else that went wrong. Were you perhaps able to see the actual metric with the reason?

Reply:
I am not sure what went wrong, but I did see the metric with the reason fine and tried variations of the query above.

Nevertheless, let's go with the simpler solution for now then.